atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/ResilientLockWatchProxy.java

```diff
@@ -14,58 +14,66 @@
+@ThreadSafe
+public final class ResilientLockWatchProxy<T> extends AbstractInvocationHandler {
+    private static final Logger log = LoggerFactory.getLogger(ResilientLockWatchProxy.class);
 
-    private static final Logger log = LoggerFactory.getLogger(ResilientLockWatchEventCache.class);
-
-    static LockWatchEventCache newProxyInstance(
+    public static LockWatchEventCache newEventCacheProxy(
             LockWatchEventCache defaultCache, LockWatchEventCache fallbackCache, MetricsManager metricsManager) {
         return (LockWatchEventCache) Proxy.newProxyInstance(
                 LockWatchEventCache.class.getClassLoader(),
                 new Class<?>[] {LockWatchEventCache.class},
-                new ResilientLockWatchEventCache(defaultCache, fallbackCache, metricsManager));
+                new ResilientLockWatchProxy<>(defaultCache, fallbackCache, metricsManager));
+    }
+
+    public static LockWatchValueScopingCache newValueCacheProxy(
+            LockWatchValueScopingCache defaultCache,
+            LockWatchValueScopingCache fallbackCache,
+            MetricsManager metricsManager) {
+        return (LockWatchValueScopingCache) Proxy.newProxyInstance(
+                LockWatchValueScopingCache.class.getClassLoader(),
+                new Class<?>[] {LockWatchValueScopingCache.class},
+                new ResilientLockWatchProxy<>(defaultCache, fallbackCache, metricsManager));
     }
 
-    private final LockWatchEventCache fallbackCache;
+    private final T fallbackCache;
     private final Counter fallbackCacheSelectedCounter;
 
-    @GuardedBy("this")
-    private LockWatchEventCache delegate;
+    private volatile T delegate;
 
-    private ResilientLockWatchEventCache(
-            LockWatchEventCache defaultCache, LockWatchEventCache fallbackCache, MetricsManager metricsManager) {
+    private ResilientLockWatchProxy(T defaultCache, T fallbackCache, MetricsManager metricsManager) {
         this.delegate = defaultCache;
         this.fallbackCache = fallbackCache;
         this.fallbackCacheSelectedCounter =
-                metricsManager.registerOrGetCounter(ResilientLockWatchEventCache.class, "fallbackCacheSelectedCounter");
+                metricsManager.registerOrGetCounter(ResilientLockWatchProxy.class, "fallbackCacheSelectedCounter");
     }
 
     @Override
-    protected synchronized Object handleInvocation(Object proxy, Method method, Object[] args)
-            throws IllegalAccessException {
+    protected Object handleInvocation(Object proxy, Method method, Object[] args) throws IllegalAccessException {
         try {
             return method.invoke(delegate, args);
         } catch (InvocationTargetException e) {
             throw handleException(e);
         }
     }
 
-    synchronized RuntimeException handleException(InvocationTargetException rethrow) {
+    private synchronized RuntimeException handleException(InvocationTargetException rethrow) {
```

Comment:
I don't think that this is an issue being synchronised, as we probably won't be throwing that many exceptions, and when we do, it should be fine. However, we could probably drop this, and at worst we'd try and fallback twice.

Reply:
It seems fine to keep it synchronized